keps/sig-node/4680-add-resource-health-to-pod-status/README.md

```diff
@@ -120,31 +120,77 @@ The proposal is to introduce an additional field:
+
+const (
+	ResourceHealthStatusHealthy   ResourceHealthStatus = "Healthy"
+	ResourceHealthStatusUnhealthy ResourceHealthStatus = "Unhealthy"
+	ResourceHealthStatusUnknown   ResourceHealthStatus = "Unknown"
+)
+
 // ResourceID is calculated based on the source of this resource health information.
 // For DevicePlugin:
-//   deviceplugin:DeviceID, where DeviceID is from the Device structure of DevicePlugin's ListAndWatchResponse type: https://github.com/kubernetes/kubernetes/blob/eda1c780543a27c078450e2f17d674471e00f494/staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1alpha/api.proto#L61-L73
+//
+//	deviceplugin:DeviceID, where DeviceID is from the Device structure of DevicePlugin's ListAndWatchResponse type: https://github.com/kubernetes/kubernetes/blob/eda1c780543a27c078450e2f17d674471e00f494/staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1alpha/api.proto#L61-L73
+//
 // DevicePlugin ID is usually a constant for the lifetime of a Node and typically can be used to uniquely identify the device on the node.
 // For DRA:
-//   dra:<driver name>/<pool name>/<device name>: such a device can be looked up in the information published by that DRA driver to learn more about it. It is designed to be globally unique in a cluster.
+//
+//	dra:<driver name>/<pool name>/<device name>: such a device can be looked up in the information published by that DRA driver to learn more about it. It is designed to be globally unique in a cluster.
 type ResourceID string
 
+// ResourceHealth represents the health of a resource. It has the latest device health information.
+// This is a part of KEP https://kep.k8s.io/4680 and historical health changes are planned to be added in future iterations of a KEP.
 type ResourceHealth struct {
-    // List of conditions with the transition times
-    Conditions []ResourceHealthCondition
+	// ResourceID is the unique identifier of the resource. See the ResourceID type for more information.
+	ResourceID ResourceID `json:"resourceID" protobuf:"bytes,1,opt,name=resourceID"`
+	// Health of the resource.
+	// can be one of:
+	//  - Healthy: operates as normal
+	//  - Unhealthy: reported unhealthy. We consider this a temporary health issue
+	//               since we do not have a mechanism today to distinguish
+	//               temporary and permanent issues.
+	//  - Unknown: The status cannot be determined.
+	//             For example, Device Plugin got unregistered and hasn't been re-registered since.
+	//
+	// In future we may want to introduce the PermanentlyUnhealthy Status.
+	Health ResourceHealthStatus `json:"health,omitempty" protobuf:"bytes,2,name=health"`
+}
+```
+
+In alpha2 we will enhance the structure `ResourceHealth` to include past conditions when devices were
+moving to unhealthy and back. The structure will be following the structure of other conditions in Pod Status:
+
+```
+type ResourceHealth struct {
+  ...
+  // List of conditions with the transition times
+  Conditions []ResourceHealthCondition
```

Comment:
@SergeyKanzhelev the question is whether `Conditions` belongs only in #4817 not here. One consideration is how we capture troubleshooting information that is per-pod-per-device (ResourceClaim status would be per-device only). If the ResourceClaim allocates some device, and it is then shared among multiple pods, I think here would be the place to capture that info. For example, suppose you successfully mount the device in one pod's container but not in another pod's container - we would want that to show up in the right place.

Reply:
Making this a slice seem unusual. I get that this is meant to capture historic transitions, but that's not how other conditions are used.

I agree that the API server should not be used to accumulate information over time. It opens a can of worms (how many entries can be stored before the Pod resource becomes too large?).

If it isn't a slice, then the information is mostly the same as in  ResourceHealth itself, just formatted like a condition.

BTW, there is https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#Condition. I think that's supposed to be used instead of defining your own custom struct now.

Reply:
removed this section